### ISSUE TITLE
incompatible types in assignment not found when type of variable is implicit

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#2823]
**Describe the bug**
```python
b = {"b": 2}
b = [5, 6]
```
pyright doesn't say that this is an incompatible assignment.  mypy _does_ flag it.  This was surprising to us.

Note if you do declare a type for 'b' as in 
```python
from typing import Dict
b: Dict[str, int] = {"b": 2}
b = [5, 6]
```
then pyright does pick that up.

**VS Code extension or command-line**
VS Code pylance v2022.1.0

**Comment by erictraut** [comment]
This is intentional behavior. The type of `b` is inferred from its assignments. In this case, its type is a union of a dict and a list. Mypy's behavior leads to significant false positives and the need for extra explicit type annotations. Mypy's issue tracker is littered with reports of this. Pyright's behavior is still type safe but much more convenient.

If your intent is to limit a variable to a specific type, you can explicitly declare its type as you've noted.

**Comment by erictraut** [comment]
For more details about pyright's type inference, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#multi-assignment-type-inference).
